doc: Make the Kconfig reference a separate documentation set

ulfalizer opened the pull request:
Instead of having a separate Kconfig reference page in each
documentation set (nRF, nrfxlib, MCUBoot, and Zephyr), have a single
shared Kconfig documentation set, which can be selected in the
documentation switcher on the bottom left.

All Kconfig symbols now get pulled in via the main Zephyr repo, which
includes nrf, nrfxlib, and mcuboot as modules. This makes it possible to
click through to symbols defined in other repositories. Previously,
symbols defined in other repositories became dead links.

The new --modules flag to genrest.py is used to generate separate index
pages for each repository.

Intersphinx is used to link :option:`CONFIG_FOO` references in all
documentation sets to the new Kconfig documentation set.

Also update the Zephyr repository to get the fw-nrfconnect-zephyr
documentation build change in
NordicPlayground/fw-nrfconnect-zephyr#213 in,
which is required by this PR.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## pylint (internal test error)

```
pylint not found. Check that the directory it's in is listed in the PATH environment variable.
```

## Identity/Emails issues

e4cff0f101f11408afff00d14191de7c69472fca: author email (Ulf Magnusson <ulfalizer@gmail.com>) needs to match one of the signed-off-by entries.



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

ulfalizer commented:
@ru-fu @carlescufi @others
Feel free to push to this PR btw. Could do a `Co-authored-by:` thing.

pdunaj commented:
@MarekPieta

ru-fu commented:
I built the output, and the structure looks great. There are some errors when building though.

The kconfig build gives me hundreds of:
```
<rst_epilog>:2: WARNING: Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: 'kconfig/rst/substitutions.txt'.
```

The zephyr build gives me a:
`C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\zephyr\rst\doc\reference\index.rst:6: WARNING: toctree contains reference to nonexisting document 'reference/kconfig/index'`

Some other things I spotted:
- Can we call the main target kconfig instead of kconfig-html (or add a second target) to be consistent with the other targets?
- In the output of each option, there is a path with, for example, `<nRF>\doc\_build\Kconfig.modules:5`. It's great that the path doesn't have the full path anymore (at least I think it doesn't?). Could we get rid of the doc\_build as well in some way?
- <nRF> should be <nrf> because that is the default folder name.
- Not sure if that is possible to do (maybe with JavaScript?), but it would be great to have a button/link to go back to the previous page for each Kconfig option. A typical use case will be that users click on a link to an option in some conceptual documentation, and then they will end up somewhere completely different. Of course they can just use their browser's back button, but giving them a link in addition would improve usability.

ru-fu commented:
For the index-* pages, do they all use the same text, or can we have different text for each?

ulfalizer commented:
> The kconfig build gives me hundreds of:
> 
> ```
> <rst_epilog>:2: WARNING: Problems with "include" directive path:
> InputError: [Errno 2] No such file or directory: 'kconfig/rst/substitutions.txt'.
> ```

Fixed. `kconfig/conf.py` is mostly copy-pasted from `zephyr/conf.py`, and there was this thing that caused it:

```
rst_epilog = """
.. include:: /substitutions.txt
"""
```

Removed it.

> The zephyr build gives me a:
> `C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\zephyr\rst\doc\reference\index.rst:6: WARNING: toctree contains reference to nonexisting document 'reference/kconfig/index'`

Fixed, in https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/213. Hadn't updated `doc/reference/index.rst` to not include `kconfig/index.rst`. It doesn't exist anymore.

> * Can we call the main target kconfig instead of kconfig-html (or add a second target) to be consistent with the other targets?

It clashes with the `kconfig` target from the included `CMakeLists.txt` from the Zephyr repository, so won't work.

Could rename it to something like `kconfig-reference` maybe. What do you think?

> * In the output of each option, there is a path with, for example, `<nRF>\doc\_build\Kconfig.modules:5`. It's great that the path doesn't have the full path anymore (at least I think it doesn't?). Could we get rid of the doc_build as well in some way?

It's because `Kconfig.modules` gets dynamically generated during the build. It's used to include the Kconfig files from the modules.

The output isn't wrong, though it gets a bit ugly there. I wonder if hiding it might make things more confusing though. It's showing how it actually works there.

> * Not sure if that is possible to do (maybe with JavaScript?), but it would be great to have a button/link to go back to the previous page for each Kconfig option. A typical use case will be that users click on a link to an option in some conceptual documentation, and then they will end up somewhere completely different. Of course they can just use their browser's back button, but giving them a link in addition would improve usability.

> For the index-* pages, do they all use the same text, or can we have different text for each?

The text is hardcoded in `genrest.py` at the moment, and is the same for all index pages (only the titles are different). I could make it configurable later on though. This PR is already getting pretty large and unwieldy though, so maybe it could be done separately.

Maybe it could be another option to `--modules`, like `--modules Foo:foo:path/to/foo:foo-index-text.rst`, where `foo-index-text.rst` holds the custom index page contents.

ru-fu commented:
> Maybe it could be another option to `--modules`, like `--modules Foo:foo:path/to/foo:foo-index-text.rst`, where `foo-index-text.rst` holds the custom index page contents.

That would be great, because then we could override the text for the Zephyr options as well for our repo.

For the target, it doesn't really matter what we use if we cannot use just "kconfig". It will be inconsistent with the other targets in any case. :(

ulfalizer commented:
Cleaned up `kconfig/conf.py`. Shouldn't have any redundant stuff in it now.

ru-fu commented:
I tried building again. But when I build kconfig-html from a clean build folder, I get the following error:
```
[1/2] cmd.exe /C "cd /D C:\Users\rufu\...C:/Users/rufu/Desktop/GIT/ncs/nrfxlib"
FAILED: CMakeFiles/kconfig-content
cmd.exe /C "cd /D C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build && "C:\Program Fi
les\CMake\bin\cmake.exe" -E make_directory C:/Users/rufu/Desktop/GIT/ncs/nrf/doc
/_build/kconfig/rst && "C:\Program Files\CMake\bin\cmake.exe" -E env PYTHONPATH=
"C:\Users\rufu\Desktop\GIT\ncs\zephyr\/scripts/kconfig;" ZEPHYR_BASE=C:\Users\ru
fu\Desktop\GIT\ncs\zephyr\ srctree=C:\Users\rufu\Desktop\GIT\ncs\zephyr\ BOARD_D
IR=boards/*/*/ ARCH=* ARCH_DIR=arch/ SOC_DIR=soc/ CMAKE_BINARY_DIR=C:/Users/rufu
/Desktop/GIT/ncs/nrf/doc/_build KCONFIG_WARN_UNDEF=y KCONFIG_TURBO_MODE= KCONFIG
_DOC_MODE=1 C:/Python37/python.exe C:\Users\rufu\Desktop\GIT\ncs\zephyr\/doc/scr
ipts/genrest.py C:/Users/rufu/Desktop/GIT/ncs/nrf/doc/_build/kconfig/rst --non-m
odule-title Zephyr --modules nRF:nrf:C:/Users/rufu/Desktop/GIT/ncs/nrf nrfxlib:n
rfxlib:C:/Users/rufu/Desktop/GIT/ncs/nrfxlib"
usage: C:\Users\rufu\Desktop\GIT\ncs\zephyr\/doc/scripts/genrest.py <Kconfig> <o
utput directory>
ninja: build stopped: subcommand failed.
```

So I tried building zephyr, but that gave me:
```
[4/5] Generating HTML documentation wi...rs/rufu/Desktop/GIT/ncs/nrf/doc/zephyr
Running Sphinx v1.7.5

Configuration error:
The configuration file (or one of the modules it imports) called sys.exit()
FAILED: zephyr/CMakeFiles/html
```
Building nrf, nrfxlib, and mcuboot works (with warnings for missing references, of course).

All on Windows, after checking out your branch and running west update.

ulfalizer commented:
> I tried building again. But when I build kconfig-html from a clean build folder, I get the following error:

Ohh, sorry, looks like a `west.yml` change got lost. I'll add it back while rebasing.

This PR needs to be paired with the `genconfig.py` update in https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/213.

ulfalizer commented:
Rebased and (hopefully) working now.

ulfalizer commented:
@ru-fu @carlescufi 
Could be merged now once https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/213 gets in.

@ru-fu 
The new flags in https://github.com/zephyrproject-rtos/zephyr/pull/19840 (which is also in nRF's Zephyr now) can be used to customize the text on index pages. Ask if you have any questions.

tejlmand commented:
@ulfalizer could you update `west.yml` to point to correct SHA:
https://github.com/NordicPlayground/fw-nrfconnect-zephyr/commit/943a8651224470b6594e992fd2c58a454540796a

ulfalizer commented:
Rebased with the updated west.yml now.

ulfalizer commented:
@carlescufi 
Could remove the DNM label. Should be ready.

ulfalizer commented:
@ru-fu @carlescufi @SebastianBoe 
Any idea how it failed now? The `Details` links seem to be internal.

ru-fu commented:
`Archiving artifacts

‘_build/*/*.warnings’ doesn’t match anything: ‘_build’ exists but not ‘_build/*/*.warnings’

No artifacts found that match the file pattern "_build/*/*.warnings". Configuration error?`

Have you moved/removed the warnings files?

ulfalizer commented:
> ```
> Archiving artifacts
> 
> ‘_build/*/*.warnings’ doesn’t match anything: ‘_build’ exists but not ‘_build/*/*.warnings’
> 
> No artifacts found that match the file pattern "_build/*/*.warnings". Configuration error?
> ```
> 
> 
> Have you moved/removed the warnings files?

Not intentionally at least. Not sure what's going on there.

I realized that doing the Zephyr repository update in a separate PR probably breaks stuff. Added it to this PR, as they need to updated as a unit. Let's see if that helps...

ulfalizer commented:
Nope... still broken.

Know where I can see the error log?

SebastianBoe commented:
Hi @ru-fu , we are struggling. We have a commit in Zephyr downstream that is depends on this PR. So we need to get this PR in.

Are you able to help me reproduce / debug the DOC failure in this PR?

ulfalizer commented:
Hmm... wonder why it's frozen this time.

ulfalizer commented:
@carlescufi 
Got any logs for most recent failure?

ru-fu commented:
Here's the current error from CI:
```
WARNING: failed to reach any of the inventories with the following issues (this one is for nrf, the same error occurs for mcuboot and nrfxlib):

WARNING: intersphinx inventory '../../../../html/kconfig/objects.inv' not fetchable due to <class 'FileNotFoundError'>: [Errno 2] No such file or directory: '/jenkins_workspace/workspace/test-fw-nrfconnect-nrf_documentation/master/_build/nrf/rst/doc/nrf/../../../../html/kconfig/objects.inv'
```

I tried to build locally, but there seems to be a path issue on Windows, since I get:
```
error: failed to open index description file '/Users/rufu/Desktop/GIT/ncs/nrf':
[Errno 13] Permission denied: '/Users/rufu/Desktop/GIT/ncs/nrf'
```
Not sure what an index description file is, but it should probably be c:/Users/...

ulfalizer commented:
@ru-fu 
Think the issue is that CI needs to be updated to build the Kconfig documentation set first as well.

Don't think that code is online though.

ru-fu commented:
How about the Windows build issue?

ulfalizer commented:
@ru-fu 
Not sure how that could happen. That error is generated by `doc/scripts/genrest.py` in Zephyr if a path to an `.rst` file with a custom description text for an index page is passed but can't be opened. No custom description is passed here though.

Sure you're running the right version of the Zephyr repository?

Maybe we could until https://github.com/zephyrproject-rtos/zephyr/pull/20322 is merged. It improves the output some more, and simplifies `genrest.py`.

ulfalizer commented:
Rebased to fix merge conflict with commit 4df38c48befc ("doc: update mcuboot documentation").

@ru-fu 
Could you double-check the Windows build thing re. the Zephyr version? Could dig some more if you still get it, but no idea how it'd happen.

This version should be mergeable besides that. https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/234 needs to go in at the same time.

ulfalizer commented:
Windows spaces-in-`--modules`-arguments issue in https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/234 fixed now, with [`VERBATIM`](https://cmake.org/cmake/help/v3.16/command/add_custom_target.html) and proper quoting.

Also updated this PR to use `VERBATIM` to possibly save some later pain, though it's not needed at the moment (no spaces in arguments).

ulfalizer commented:
@ru-fu @carlescufi 

Tracked down why building the Kconfig docs takes several hours. The root cause
is the top-level `index.rst` being super large now that it lists all symbols.

Happens like this:

1. `nrf/doc/kconfig/conf.py` uses `themes/kconfig/theme.conf`, which
   inherits from `themes/nrf/theme.conf`, which uses the Sphinx-bundled
   `sphinx_rtd_theme`.

2. `sphinx_rtd_theme/layout.html` has this, which is used for the `<div>` with
   the documentation navigation menu on the left, which is based on the
   top-level toctree:

   ```html
   <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
     {% block menu %}
       {#
         The singlehtml builder doesn't handle this toctree call when the
         toctree is empty. Skip building this for now.
       #}
       {% if 'singlehtml' not in builder %}
         {% set global_toc = toctree(maxdepth=theme_navigation_depth|int,
                                     collapse=theme_collapse_navigation|tobool,
                                     includehidden=theme_includehidden|tobool,
                                     titles_only=theme_titles_only|tobool) %}
       {% endif %}
       {% if global_toc %}
         {{ global_toc }}
       {% else %}
         <!-- Local TOC -->
         <div class="local-toc">{{ toc }}</div>
       {% endif %}
     {% endblock %}
   </div>
   ```

3. That ends up unpickling the doctree for the top-level index page, like this:

   ```
   ...
   File "/usr/local/lib/python3.7/dist-packages/sphinx_rtd_theme/layout.html", line 345, in block_menu
   File "/usr/local/lib/python3.7/dist-packages/jinja2/sandbox.py", line 440, in call
     return __context.call(__obj, *args, **kwargs)
   File "/usr/local/lib/python3.7/dist-packages/jinja2/runtime.py", line 262, in call
     return __obj(*args, **kwargs)
   File "/usr/local/lib/python3.7/dist-packages/sphinx/builders/html.py", line 1003, in <lambda>
     ctx['toctree'] = lambda **kw: self._get_local_toctree(pagename, **kw)
   File "/usr/local/lib/python3.7/dist-packages/sphinx/builders/html.py", line 898, in _get_local_toctree
     docname, self, collapse, **kwds))['fragment']
   File "/usr/local/lib/python3.7/dist-packages/sphinx/environment/adapters/toctree.py", line 308, in get_toctree_for
     doctree = self.env.get_doctree(self.env.config.master_doc)
   File "/usr/local/lib/python3.7/dist-packages/sphinx/environment/__init__.py", line 783, in get_doctree
     doctree = pickle.load(f)
   ``` 

   The unpickling is super slow, since the top-level page is so large.

Removing `html_theme = "kconfig"` from `nrf/doc/kconfig/conf.py` speeds things
up 10x+.

The only workaround I've found so far is to make the top-level page smaller,
e.g. by putting the list of all symbols on a separate page. The real fix might
be to make Sphinx cache the toctree of the top-level page or the like.

ulfalizer commented:
Reported the issue on the Sphinx bug tracker too: https://github.com/sphinx-doc/sphinx/issues/6909

ru-fu commented:
I think it's actually a good idea to put the list of all symbols on a subpage. 
There are links to the "filtered" lists for the subsystems there, and my first question when I saw those was "and where is the link to the full list?!". Which is not surprising, since humans don't read text, they just scan through it. ;-)
So having "All configuration options" as additional link is more user-friendly than having them on the index page. It will also make them show up in the navigation on the left.

ulfalizer commented:
> I think it's actually a good idea to put the list of all symbols on a subpage.
> There are links to the "filtered" lists for the subsystems there, and my first question when I saw those was "and where is the link to the full list?!". Which is not surprising, since humans don't read text, they just scan through it. ;-)
> So having "All configuration options" as additional link is more user-friendly than having them on the index page. It will also make them show up in the navigation on the left.

Alright, gonna add an option for it. @dbkinder wanted all symbols on the top-level index page (and I thought it was nice in Zephyr at least), so I feel like I'm being torn in different directions here. ;)

ulfalizer commented:
Updated to use the new `--separate-all-index` option from https://github.com/zephyrproject-rtos/zephyr/pull/21272 now.

I also realized that the Kconfig reference can be built in parallel, since it doesn't use Breathe, so added `-j auto` to the `sphinx-build` command. The Kconfig reference now builds in around two minutes on my system.

dbkinder commented:
I'm fine with having an "all symbols" page (instead of being on the main page) to avoid problems with Zephyr doc build you've found.  So feel free to bring these changes as appropriate to the Zephyr master branch :)

ulfalizer commented:
Heh... didn't realize non-Nordic people can see this repo too.

Bit silly that I talked about an "internal" project in that Sphinx bug report then too. Derp...

ulfalizer commented:
Needs this documentation build update too: https://projecttools.nordicsemi.no/bitbucket/projects/NCS-TEST/repos/test-fw-nrfconnect-nrf_doc/pull-requests/23/overview

Should go in at the same time as this PR.

ru-fu commented:
Finally got around to test this again. The kconfig build still takes a long time, but it's quicker than it was (didn't time it, but about 15 minutes).

I got a few errors when building - seems like there are a few problems left with the integration:
**For nrfxlib**:
`C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\nrfxlib\rst\README.rst:15: WARNING: toctree contains reference to nonexisting document 'kconfig/index'`
**For mcuboot**:
`intersphinx inventory '..\\..\\html\\kconfig\\objects.inv' not fetchable due to <class 'FileNotFoundError'>: [Errno 2] No such file or directory: 'C:\\Users\\rufu\\Desktop\\GIT\\ncs\\nrf\\doc\\_build\\mcuboot\\rst\\doc\\mcuboot\\..\\..\\html\\kconfig\\objects.inv'`
**For nrf**:
```
C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\nrf\rst\doc\nrf\samples\bluetooth\peripheral_uart\README.rst:62: WARNING: unknown option: zephyr:CONFIG_BT_DEVICE_NAME
C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\nrf\rst\doc\nrf\ug_nrf5340.rst:78: WARNING: unknown option: zephyr:CONFIG_BT_MAX_CONN
```

Otherwise the output looks good. :-) Hope we can get this in soon!

ulfalizer commented:
> Finally got around to test this again. The kconfig build still takes a long time, but it's quicker than it was (didn't time it, but about 15 minutes).
> 
> I got a few errors when building - seems like there are a few problems left with the integration:
> **For nrfxlib**:
> `C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\nrfxlib\rst\README.rst:15: WARNING: toctree contains reference to nonexisting document 'kconfig/index'`
> **For mcuboot**:
> `intersphinx inventory '..\\..\\html\\kconfig\\objects.inv' not fetchable due to <class 'FileNotFoundError'>: [Errno 2] No such file or directory: 'C:\\Users\\rufu\\Desktop\\GIT\\ncs\\nrf\\doc\\_build\\mcuboot\\rst\\doc\\mcuboot\\..\\..\\html\\kconfig\\objects.inv'`
> **For nrf**:
> 
> ```
> C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\nrf\rst\doc\nrf\samples\bluetooth\peripheral_uart\README.rst:62: WARNING: unknown option: zephyr:CONFIG_BT_DEVICE_NAME
> C:\Users\rufu\Desktop\GIT\ncs\nrf\doc\_build\nrf\rst\doc\nrf\ug_nrf5340.rst:78: WARNING: unknown option: zephyr:CONFIG_BT_MAX_CONN
> ```
> 
> Otherwise the output looks good. :-) Hope we can get this in soon!

Fixed the issues for mcuboot and nrf: `MCUBOOT_RST_SRC` being set to the wrong value, and some ``:option:`zephyr:CONFIG_FOO` `` (changed them too just ``:option:`CONFIG_FOO` ``. intersphinx figures it out).

Will look at the nrfxlib one. Needs a patch in that repo I think.

ulfalizer commented:
Fix for nrfxlib warning: https://github.com/NordicPlayground/nrfxlib/pull/138

Should merge that one at the same time as this one, or the link to the old separated reference will disappear for nrfxlib.

ulfalizer commented:
Double-checked the output, and it seems fine. Can click `CONFIG_*` links in all the documentation sets, etc.

Merging this will be a bit awkward. Think this should work, done roughly at the same time:

1. Merge https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/234 and https://github.com/NordicPlayground/nrfxlib/pull/138

2. I'll update this PR with the SHAs. Then it could be merged.

Hopefully I didn't forget anything.

ulfalizer commented:
Hope rebasing nrfxlib fixes the Kconfig issues...

ulfalizer commented:
Had forgotten that https://projecttools.nordicsemi.no/bitbucket/projects/NCS-TEST/repos/test-fw-nrfconnect-nrf_doc/pull-requests/23 needs to be merged too, to build the Kconfig docs in CI. Added now.

ulfalizer commented:
Turned the Zephyr repository into a separate module too, as suggested by @ru-fu:

![Skärmbild från 2020-01-31 09-35-50](https://user-images.githubusercontent.com/154768/73524326-2ab52680-440d-11ea-8739-2db76243174f.png)

Also means paths get a `<Zephyr>` prefix:

![Skärmbild från 2020-01-31 09-30-20](https://user-images.githubusercontent.com/154768/73524363-402a5080-440d-11ea-9ef2-1ce5eebd1b3b.png)

Seems to work fine.

ru-fu commented on doc/kconfig/conf.py:
```diff
@@ -0,0 +1,62 @@
+copyright = '2019, Nordic Semiconductor'
+author = 'Nordic Semiconductor'
+
+version = '1.0.0'  # Dummy version
```
How do we do the versioning here? Can we just keep it unversioned, or would that be confusing?

ulfalizer replied:
Not sure if it can be left out, so I just put in some version.

Does it show up in some place now? [Looks like](http://www.sphinx-doc.org/en/master/usage/configuration.html#confval-version) it's only used to replace `|version|`.

ulfalizer replied:
Hmm... maybe it's used in other places too.

Could you try removing it and see what happens? Don't have a handy doc build environment available atm.

ru-fu replied:
```suggestion
version = '&nbsp;'  # Dummy version
```

Replace with non-breakable space to get rid of the version number, but keep the spacing the same as for the other doc sets.

ru-fu commented on doc/kconfig/conf.py:
```diff
@@ -0,0 +1,62 @@
+master_doc = 'index'
+
+# General information about the project.
+project = 'Kconfig Reference'
```
```suggestion
project = 'Kconfig reference'
```

ulfalizer replied:
Fixed

ru-fu commented on doc/kconfig/conf.py:
```diff
@@ -0,0 +1,62 @@
+
+# The name for this set of Sphinx documents.  If None, it defaults to
+# "<project> v<release> documentation".
+html_title = "Kconfig Reference"
```
```suggestion
html_title = "Kconfig reference"
```

ulfalizer replied:
Fixed

ru-fu commented on doc/themes/kconfig/versions.html:
```diff
@@ -0,0 +1,21 @@
+{# Add rst-badge after rst-versions for small badge style. #}
+  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
+    <span class="rst-current-version" data-toggle="rst-current-version">
+      <span class="fa fa-book">Kconfig Reference</span>
```
```suggestion
      <span class="fa fa-book"> Kconfig Reference</span>
```

ulfalizer replied:
Guessing it was meant to lowercase the `R`, so fixed that. :)

ru-fu commented:
> The main concern I have is that if you search for a Zephyr kconfig option from within the zephyr documentation using the search bar at the top left, you get no results. This makes sense technically, but it's a difference from how mainline documentation works which may be surprising.
> 
> @ru-fu do you think this difference needs special mention within the NCS documentation?

We can mention it in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/documentation.html , but I don't think this will prevent any confusion ...

Could we somehow keep the link to "Configuration options" in the navigation, but change it to point to the respective page in the new docset? This won't help for search, but might still make it easier to find the correct options.

Another thing to look into would be if we can somehow combine all the different search functions (ideally so users can choose if they want to search only the current doc set or all of them). But I have no idea if that is technically feasible.

mbolivar commented:
> Could we somehow keep the link to "Configuration options" in the navigation, but change it to point to the respective page in the new docset? This won't help for search, but might still make it easier to find the correct options.

Sounds like the sort of thing that could be done with a redirect.



> Another thing to look into would be if we can somehow combine all the different search functions (ideally so users can choose if they want to search only the current doc set or all of them). But I have no idea if that is technically feasible.

No idea either, but that sounds nontrivial.

ru-fu commented on doc/themes/kconfig/versions.html:
```diff
@@ -0,0 +1,21 @@
+{# Add rst-badge after rst-versions for small badge style. #}
+  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
+    <span class="rst-current-version" data-toggle="rst-current-version">
+      <span class="fa fa-book">Kconfig reference</span>
```
```suggestion
      <span class="fa fa-book"> Kconfig reference</span>
```
The main point was actually the added space - it will look strange otherwise.

ulfalizer replied:
Fixed

ru-fu commented on doc/kconfig/conf.py:
```diff
@@ -0,0 +1,62 @@
+copyright = '2019, Nordic Semiconductor'
+author = 'Nordic Semiconductor'
+
+version = '1.0.0'  # Dummy version
```
```suggestion
version = '&nbsp;'  # Dummy version
```

Replace with non-breakable space to get rid of the version number, but keep the spacing the same as for the other doc sets.

ulfalizer replied:
Fixed

thst-nordic commented:
I linked the patch in test-fw-nrfconnnect-doc/PR-23 to this build.  
need to fix the conflicts before it will get to this stage.

carlescufi commented:
Ready for merge as soon as CI is green
